### ISSUE TITLE
Feature: Bloomberg's Memray

### ISSUE THREAD

**Issue by Taz33m** [barbarakotlan/MarketMind#88]
## Objective

Add an optional [Bloomberg Memray](https://github.com/bloomberg/memray) profiling workflow to MarketMind so we can investigate backend memory growth, leaks, and allocation hotspots in a structured way without affecting normal runtime behavior.

## Problem Statement

MarketMind has several Python-heavy surfaces where memory usage can become expensive or unclear over time, especially during debugging and performance work:

- the long-running backend API process
- data-heavy market fetch / processing paths
- ML and benchmark workflows
- selective evaluation / training jobs using pandas, numpy, scikit-learn, and xgboost

Right now we do not have a first-class memory profiling workflow for these paths.

That makes it harder to answer questions like:

- Is backend RSS growth coming from Python objects or native allocations?
- Is a repeated route flow leaking memory?
- Which functions allocate the most memory during a train/eval/benchmark job?
- Are large memory spikes temporary or persistent?

Without a dedicated profiler, we are mostly guessing from symptoms instead of inspecting allocations directly.

## Why Memray

[Memray](https://github.com/bloomberg/memray) is a strong fit for MarketMind because it:

- profiles Python memory allocations with full call-stack tracing
- can track native allocations from C/C++ extensions
- works well with pandas / numpy / xgboost style workloads
- supports both one-off script profiling and long-running process analysis
- can generate summary, stats, and flamegraph reports
- runs on macOS and Linux, which matches our development environments

This makes it more useful for MarketMind than a Python-only memory tool that cannot see native allocation behavior.

## Scope

This feature should add a **developer-facing, optional** memory profiling workflow for the Python backend.

### In Scope

- add Memray as an optional backend development dependency
- provide a documented workflow for profiling:
  - backend API startup / route exercise flows
  - selective train/eval/benchmark jobs
- provide simple helper scripts for:
  - running the backend under Memray
  - running `backend.selective_prediction_global` under Memray
  - generating human-readable reports from captures
- store generated profiling artifacts outside normal source control
- document when Memray should and should not be used

### Out of Scope

- adding Memray to production runtime or deploy paths
- automatically running Memray on every backend startup
- changing backend behavior, routes, or API contracts
- turning this into a mandatory CI gate in the first iteration
- frontend memory profiling

### 1. Optional dependency path

Add a backend development dependency path that includes Memray without changing normal runtime installation.

Example shape:

- `backend/requirements-dev.txt`

This keeps profiling tooling separate from the main backend dependency set.

### 2. Backend API profiling workflow

Add a helper script to run the Flask backend under Memray from process start.

Purpose:

- profile startup allocations
- reproduce leaks or growth during repeated route usage
- inspect memory behavior during local debugging sessions

Expected behavior:

- launch `backend/api.py` under `memray run`
- default to writing captures into a dedicated profiling output directory
- allow overriding Memray CLI flags when needed
- prefer `--native` by default for realistic pandas/numpy/xgboost visibility

### 3. Selective / benchmark job profiling workflow

Add a helper script to run `backend.selective_prediction_global` under Memray.

Purpose:

- inspect train/eval/benchmark memory spikes
- identify allocation hotspots in data prep, model training, and benchmarking
- give us a repeatable profiling path for the heaviest current backend jobs

Expected behavior:

- forward command arguments through to `python -m backend.selective_prediction_global`
- support train/eval/benchmark-label-strategies flows
- default to writing captures into the same profiling output directory
- prefer `--native` by default

### 4. Report generation workflow

Add a helper script to generate reports from a Memray capture.

Expected outputs:

- summary report
- stats report
- flamegraph HTML

The reporting flow should work fully offline/local and should not rely on assets that make local generation flaky.

### 5. Documentation

Add a short backend profiling README covering:

- what Memray is
- why it is useful in MarketMind
- install steps
- recommended workflows
- example commands
- warnings about attach mode
- guidance on when to use it vs when not to use it

## Deliverables

- optional Memray dependency path for backend development
- helper script for profiling backend API startup/runtime
- helper script for profiling selective/benchmark jobs
- helper script for generating reports from captures
- profiling README with usage examples
- Git ignore rule for generated Memray output

## Acceptance Criteria

- Memray can be installed cleanly into the backend virtual environment
- a developer can profile the backend API with one command
- a developer can profile a selective benchmark/train/eval job with one command
- a developer can generate summary, stats, and flamegraph reports from a capture
- generated profiling artifacts are not committed accidentally
- no backend runtime behavior changes for users who are not using Memray
- no API route changes
- no production dependency or startup coupling

### Profile backend API

```bash
backend/profiling/memray/run_api_under_memray.sh
Profile a selective job
backend/profiling/memray/run_selective_global_under_memray.sh -- \
  eval \
  --tickers AAPL,MSFT,NVDA,SPY \
  --global-root backend/model_artifacts/selective_global/v2
Generate reports
backend/profiling/memray/report_memray_capture.sh \
  backend/profiling/memray/output/<capture>.bin
